scrapy_zyte_api/handler.py

```diff
@@ -42,24 +45,33 @@ def __init__(
         if not settings.getbool("ZYTE_API_ENABLED", True):
             raise NotConfigured
         if not client:
-            try:
-                client = AsyncClient(
-                    # To allow users to have a key defined in Scrapy settings
-                    # and in a environment variable, and be able to cause the
-                    # environment variable to be used instead of the setting by
-                    # overriding the setting on the command-line to be an empty
-                    # string, we do not support setting empty string keys
-                    # through settings.
-                    api_key=settings.get("ZYTE_API_KEY") or None,
-                    api_url=settings.get("ZYTE_API_URL") or API_URL,
-                    n_conn=settings.getint("CONCURRENT_REQUESTS"),
-                )
-            except NoApiKey:
-                logger.warning(
-                    "'ZYTE_API_KEY' must be set in the spider settings or env var "
-                    "in order for ScrapyZyteAPIDownloadHandler to work."
-                )
-                raise NotConfigured
+            client_key = (
+                settings.get("ZYTE_API_KEY") or None,
+                settings.get("ZYTE_API_URL") or API_URL,
+                settings.getint("CONCURRENT_REQUESTS"),
+            )
+            if client_key in _CLIENT_CACHE:
+                client = _CLIENT_CACHE[client_key]
+            else:
+                try:
+                    client = AsyncClient(
+                        # To allow users to have a key defined in Scrapy settings
+                        # and in a environment variable, and be able to cause the
+                        # environment variable to be used instead of the setting by
+                        # overriding the setting on the command-line to be an empty
+                        # string, we do not support setting empty string keys
+                        # through settings.
+                        api_key=client_key[0],
+                        api_url=client_key[1],
+                        n_conn=client_key[2],
+                    )
+                except NoApiKey:
+                    logger.warning(
+                        "'ZYTE_API_KEY' must be set in the spider settings or env var "
+                        "in order for ScrapyZyteAPIDownloadHandler to work."
+                    )
+                    raise NotConfigured
+                _CLIENT_CACHE[client_key] = client
```

Comment:
One more alternative: store the client object as an attribute of either `crawler` or `spider`. Basically, we need some storage which is shared between all components in a crawl. If we use spider, it would also allow users to define custom AsyncClient instance as a spider attribute.

Reply:
If we use cache with settings as keys, then stats are going to be shared between spiders running in the same process, if the options are the same.